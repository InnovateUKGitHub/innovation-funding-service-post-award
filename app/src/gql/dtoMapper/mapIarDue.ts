type IARClaimNode = Readonly<{
  Acc_IAR_Status__c: GQL.Value<string>;
  Acc_IARRequired__c: GQL.Value<boolean>;
  Acc_ProjectID__c: GQL.Value<string>;
}> | null;

/**
 * Calculates the periods for which IAR is due
 */
export const getIARDueOnClaimPeriods = (claimsGql: ReadonlyArray<{ node: IARClaimNode | null } | null>) =>
  claimsGql.reduce((iarPeriods: string[], cur) => {
    if (cur?.node?.Acc_IAR_Status__c?.value === "Not Received" && !!cur?.node?.Acc_IARRequired__c?.value) {
      iarPeriods.push((cur?.node?.Acc_ProjectID__c?.value ?? 0).toString());
    }
    return iarPeriods;
  }, []) ?? [];
