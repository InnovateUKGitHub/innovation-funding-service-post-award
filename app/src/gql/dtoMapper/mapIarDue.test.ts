import { getIARDueOnClaimPeriods } from "./mapIarDue";

describe("getIARDueOnClaimPeriods", () => {
  const edges = [
    {
      node: {
        RecordType: {
          DeveloperName: { value: "Total_Project_Period" },
        },
        Acc_IAR_Status__c: { value: "Not Received" },
        Acc_IARRequired__c: { value: true },
        Acc_ProjectPeriodNumber__c: { value: 1 },
      },
    },
    {
      node: {
        RecordType: {
          DeveloperName: { value: "Total_Project_Period" },
        },
        Acc_IAR_Status__c: { value: "Not Received" },
        Acc_IARRequired__c: { value: true },
        Acc_ProjectPeriodNumber__c: { value: 2 },
      },
    },
    {
      node: {
        RecordType: {
          DeveloperName: { value: "Total_Project_Period" },
        },
        Acc_IAR_Status__c: { value: "Not Received" },
        Acc_IARRequired__c: { value: false },
        Acc_ProjectPeriodNumber__c: { value: 3 },
      },
    },
    {
      node: {
        RecordType: {
          DeveloperName: { value: "Unwanted_Record_Type" },
        },
        Acc_IAR_Status__c: { value: "Received" },
        Acc_IARRequired__c: { value: true },
        Acc_ProjectPeriodNumber__c: { value: 3.5 },
      },
    },
    {
      node: {
        RecordType: {
          DeveloperName: { value: "Total_Project_Period" },
        },
        Acc_IAR_Status__c: { value: "Received" },
        Acc_IARRequired__c: { value: true },
        Acc_ProjectPeriodNumber__c: { value: 4 },
      },
    },
  ];

  it("should map to an array of projectIds for which IAR is due", () => {
    expect(getIARDueOnClaimPeriods(edges)).toMatchSnapshot();
  });
});
