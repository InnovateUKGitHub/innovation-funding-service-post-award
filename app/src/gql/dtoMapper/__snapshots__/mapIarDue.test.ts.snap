// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIARDueOnClaimPeriods should map to an array of projectIds for which IAR is due 1`] = `
[
  "1",
  "2",
]
`;
